packages/tinacms/src/tina-cms.tsx

```diff
@@ -289,23 +249,167 @@ export const TinaCMSProvider2 = ({
     <TinaCloudProvider
       branch={branch}
       clientId={clientId}
-      tinaioConfig={tinaioConfig}
+      tinaioConfig={props.tinaioConfig}
       isLocalClient={isLocalClient}
-      cmsCallback={cmsCallback}
-      mediaStore={mediaStore}
+      cmsCallback={props.cmsCallback}
+      mediaStore={props.mediaStore}
     >
-      {props.query ? (
-        <SetupHooks key={props.query} {...props} query={props.query || ''}>
-          {children}
-        </SetupHooks>
-      ) : (
-        // @ts-ignore
-        children(props)
-      )}
+      <TinaCMSProviderWithQuery {...props} query={query} />
     </TinaCloudProvider>
   )
 }
 
+const TinaCMSProviderWithQuery = ({
+  children,
+  cmsCallback,
+  mediaStore,
+  tinaioConfig,
+  ...props
+}: {
+  /** The query from getStaticProps */
+  query: string
+  /** Any variables from getStaticProps */
+  variables?: object
+  /** The `data` from getStaticProps */
+  data: object
+  /** Your React page component */
+  children: (props?: any) => React.ReactNode
+  /**
+   * The URL for the GraphQL API.
+   *
+   * When working locally, this should be http://localhost:4001/graphql.
+   *
+   * For Tina Cloud, use https://content.tinajs.io/content/my-client-id/github/my-branch
+   */
+  apiURL: string
+  /**
+   * Point to the local version of GraphQL instead of tina.io
+   * https://tina.io/docs/tinacms-context/#adding-tina-to-the-sites-frontend
+   *
+   * @deprecated use apiURL instead
+   */
+  isLocalClient?: boolean
+  /**
+   * The base branch to pull content from. Note that this is ignored for local development
+   *
+   * @deprecated use apiURL instead
+   */
+  branch?: string
+  /**
+   * Your clientID from tina.aio
+   *
+   * @deprecated use apiURL instead
+   */
+  clientId?: string
+  /** Callback if you need access to the TinaCMS instance */
+  cmsCallback?: (cms: TinaCMS) => TinaCMS
+  /** Callback if you need access to the "formify" API */
+  formifyCallback?: formifyCallback
+  /** Callback if you need access to the "document creator" API */
+  documentCreatorCallback?: Parameters<typeof useDocumentCreatorPlugin>[0]
+  /** TinaCMS media store instance */
+  mediaStore?:
+    | TinaCloudMediaStoreClass
+    | (() => Promise<TinaCloudMediaStoreClass>)
+  tinaioConfig?: TinaIOConfig
+}) => {
+  const cms = useCMS()
+
+  // Note - this will only register for legacy implementations taking in query
+  // Hopefuly things work if we call this twice
+  const [payload, isLoading] = useGraphqlForms({
+    query: (gql) => (props.query ? gql(props.query) : undefined),
+    variables: props.variables || {},
+    formify: (args) => {
+      if (props.formifyCallback) {
+        return props.formifyCallback(args, cms)
+      } else {
+        return args.createForm(args.formConfig)
+      }
+    },
+  })
+
+  useDocumentCreatorPlugin(props.documentCreatorCallback)
+  return (
+    <ErrorBoundary>
+      <TinaDataProvider formifyCallback={props.formifyCallback}>
+        {isLoading && (
```

Comment:
Fixed an existing issue here where we were unmounting/mounting children here when loading state changed